src/analysis/inner_product_space/linear_pmap.lean

```diff
@@ -0,0 +1,216 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Moritz Doll
+-/
+
+import analysis.inner_product_space.adjoint
+import topology.algebra.module.linear_pmap
+import topology.algebra.module.basic
+
+/-!
+
+# Partially defined linear operators on Hilbert spaces
+
+We will develop the basics of the theory of unbounded operators on Hilbert spaces.
+
+## Main definitions
+
+* `linear_pmap.is_formal_adjoint`: An operator `T` is a formal adjoint of `S` if for all `x` in the
+domain of `T` and `y` in the domain of `S`, we have that `⟪T x, y⟫ = ⟪x, S y⟫`.
+* `linear_pmap.adjoint`: The adjoint of a map `E →ₗ.[𝕜] F` as a map `F →ₗ.[𝕜] E`.
+
+## Main statements
+
+* `linear_pmap.adjoint_is_formal_adjoint`: The adjoint is a formal adjoint
+* `linear_pmap.is_formal_adjoint.le_adjoint`: Every formal adjoint is contained in the adjoint
+* `continuous_linear_map.to_pmap_adjoint_eq_adjoint_to_pmap_of_dense`: The adjoint on
+`continuous_linear_map` and `linear_pmap` coincide.
+
+## Notation
+
+* For `T : E →ₗ.[𝕜] F` the adjoint can be written as `T†`.
+This notation is localized in `linear_pmap`.
+
+## Implementation notes
+
+We use the junk value pattern to define the adjoint for all `linear_pmap`s. In the case that
+`T : E →ₗ.[𝕜] F` is not densely defined the adjoint `T†` is the zero map from `T.adjoint_domain` to
+`E`.
+
+## References
+
+* [J. Weidmann, *Linear Operators in Hilbert Spaces*][weidmann_linear]
+
+## Tags
+
+Unbounded operators, closed operators
+-/
+
+
+noncomputable theory
+
+open is_R_or_C
+open_locale complex_conjugate classical
+
+variables {𝕜 E F G : Type*} [is_R_or_C 𝕜]
+variables [normed_add_comm_group E] [inner_product_space 𝕜 E]
+variables [normed_add_comm_group F] [inner_product_space 𝕜 F]
+
+local notation `⟪`x`, `y`⟫` := @inner 𝕜 _ _ x y
+
+namespace linear_pmap
+
+/-- An operator `T` is a formal adjoint of `S` if for all `x` in the domain of `T` and `y` in the
+domain of `S`, we have that `⟪T x, y⟫ = ⟪x, S y⟫`. -/
+def is_formal_adjoint (T : E →ₗ.[𝕜] F) (S : F →ₗ.[𝕜] E) : Prop :=
+∀ (x : T.domain) (y : S.domain), ⟪T x, y⟫ = ⟪(x : E), S y⟫
+
+variables {T : E →ₗ.[𝕜] F} {S : F →ₗ.[𝕜] E}
+
+@[protected] lemma is_formal_adjoint.symm (h : T.is_formal_adjoint S) : S.is_formal_adjoint T :=
+λ y _, by rw [←inner_conj_symm, ←inner_conj_symm (y : F), h]
+
+variables (T)
+
+/-- The domain of the adjoint operator.
+
+This definition is needed to construct the adjoint operator and the preferred version to use is
+`T.adjoint.domain` instead of `T.adjoint_domain`. -/
+def adjoint_domain : submodule 𝕜 F :=
+{ carrier := {y | continuous ((innerₛₗ 𝕜 y).comp T.to_fun)},
+  zero_mem' := by { rw [set.mem_set_of_eq, linear_map.map_zero, linear_map.zero_comp],
+      exact continuous_zero },
+  add_mem' := λ x y hx hy, by { rw [set.mem_set_of_eq, linear_map.map_add] at *, exact hx.add hy },
+  smul_mem' := λ a x hx, by { rw [set.mem_set_of_eq, linear_map.map_smulₛₗ] at *,
+    exact hx.const_smul (conj a) } }
```

Comment:
Given this doesn't mention `T.domain`, should it just be defined for `T'` a regular `linear_map`?

Reply:
You are technically right, but I think this definition is completely uninteresting apart from the construction of the adjoint. I wanted to have that `T.adjoint.domain = T.adjoint_domain`, under your proposal it would be `T.adjoint.domain = T.to_fun.cont_inner_comp` or something similar.

Reply:
Your "the preferred version to use is `T.adjoint.domain` instead of `T.adjoint_domain`" comment suggests that you don't want people to use this definition anyway though!

Would `T.to_fun.adjoint_domain` be more appealing than `T.to_fun.cont_inner_comp`?

Reply:
Yes, I made it a definition because inlining is super messy.

Reply:
One important point is that the domain is independent of the property that the domain of `T` is dense. Otherwise it would be impossible to build sensible API for the operator. So I have to define the domain first and then build the `linear_pmap` from `adjoint_aux : T.adjoint_domain -> E` with junk value `0 : T.adjoint_domain -> E`.